### PR TITLE
TEP-0103: Skipping Reason

### PR DESCRIPTION
**Pull request by chitrangpatel**
# Changes

Add `SkippingReason` to `SkippedTasks` field of `PipelineRunStatus`.

Prior to this, users only knew that a `PipelineTask` was skipped but not
the exact reason for skipping the task leading to confusion when
debugging `Pipelines`. 

This PR addresses the [TEP](https://github.com/tektoncd/community/blob/main/teps/0103-skipping-reason.md) and [issue](https://github.com/tektoncd/pipeline/issues/4738). It adds a field `Reason` of type [`SkippingReason`](https://github.com/tektoncd/pipeline/blob/053833cb10f3829d5a366daa1f431b293dcf3285/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L42-L62) (a string alias) to the [`SkippedTasks`](https://github.com/tektoncd/pipeline/blob/053833cb10f3829d5a366daa1f431b293dcf3285/pkg/apis/pipeline/v1beta1/pipelinerun_types.go#L466-L476)field in `PipelineRunStatus`. The tests have been updated to validate skipping reason.

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in
(if there are no user facing changes, use release note "NONE")

# Release Notes
```
Add SkippingReason to SkippedTasks field of PipelineRunStatus.
```

### CODE REVIEW

**Comment by dibyom** [comment]
Closed by mistake

**Review by jerop — COMMENTED** [review]
@chitrangpatel - thank you! 

Please squash the commits into one and write a description, as described in the [commit messages standards](https://github.com/tektoncd/community/blob/main/standards.md#commits) - this could be the summary from [TEP-0103](https://github.com/tektoncd/community/blob/main/teps/0103-skipping-reason.md).

Today, we add the `when` expressions to `skippedTasks` when a `PipelineTask` is skipped regardless of the reason - this is useful for users to deduce whether the reason was its `when` expressions or not. Now that the reason is explicit, we can add the `when` expressions only when it's skipped because of `when` expressions evaluating to false. This ensures that we provide the necessary information only, and helps keep the `PipelineRun` status minimal as described in [TEP-0100](https://github.com/tektoncd/community/blob/main/teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md). This can be done as follow up work after this PR is merged. 

(cc @pritidesai)

**Comment by jerop** [comment]
Closes https://github.com/tektoncd/pipeline/issues/4738

File: docs/pipelineruns.md [hunk 1]
```diff
@@ -676,6 +676,7 @@ Conditions:
   Type:                  Succeeded
 Skipped Tasks:
   Name:       skip-this-task
+  Reason:     WhenExpressionsSkip
```
Comment (jerop) [inline docs/pipelineruns.md:679]: please update lines 666-668 as well to indicate that a Reason field will be added

Comment (chitrangpatel) [reply docs/pipelineruns.md:681]: Done.

**Comment by chitrangpatel** [comment]
Closing this PR due to merge commits affecting a squash-down. 
Opening a new PR https://github.com/tektoncd/pipeline/pull/4829 instead.

**Comment by afrittoli** [comment]
Hello @chitrangpatel and thank you for your PR. 

Opening a new PR is fine, however it has the downside of dispersing the context across multiple PRs.

For next time, if you have an issue on your local branch, you can still use the existing PR.
Once you have a working local branch, you can push it to your existing PR, for instance:

```shell
git push <fork remote name> <local branch name:existing pr branch> --force
```

In the case of this PR, the existing pr branch would be `TEP-0103-Skipping-Reason`

**Comment by chitrangpatel** [comment]
Thanks @afrittoli! I will do that next time.

**Pull request closed without merging**